include/envoy/http/filter.h

```diff
@@ -198,6 +198,22 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
    * @param trailers supplies the trailers to encode.
    */
   virtual void encodeTrailers(HeaderMapPtr&& trailers) PURE;
+
+  /**
+   * Called when the buffer for a decoder filter or any buffers the filter sends data to go over
+   * their high watermark.
+   *
+   * In the case of a filter such as the router filter, which spills into multiple buffers (codec,
```

Comment:
Having read more of the PR, I think a MD file explaining the connection + stream level design stuff, an then real examples of callback sequences on different events/overflows would be super helpful. Looking through the PR locally, it looks very solid, but I'm having a hard time connecting all the dots staring at the code, some worked examples or diagrams if you prefer would be great.